### PR TITLE
Attention with output permutation

### PR DESCRIPTION
**Pull request by rosenrodt**
Add device level function `DeviceBatchedGemmSoftmaxGemmPermute_Xdl_CShuffle`. Replacements:
- CLayout -> CPermuteNumDims_G_M_Gemm1N which is a `Sequence<>` 
- StrideC, BatchStrideC -> c_gs_ms_os_lengths, c_gs_ms_os_strides which are `std::vector<index_t>`'s

I do not implement full contraction-based solution because I have concerns about interface design of current contraction ops, mainly that `TensorSpecialization::Packed` seems broken unless having an extra input "DimOrder".

### CODE REVIEW

File: include/ck/tensor_operation/gpu/device/gemm_specialization.hpp [hunk 1]
```diff
@@ -17,6 +18,15 @@ enum struct GemmSpecialization
     MKPadding,
     NKPadding,
     MNKPadding,
+    // Gemm + Gemm
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/gemm_specialization.hpp:21]: Better use 2 separate gemm_specialization for first and second GEMM

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_softmax_gemm_permute.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                        AElementwiseOperation a_element_op,
+                        B0ElementwiseOperation b0_element_op,
+                        Acc0ElementwiseOperation acc0_element_op,
+                        B1ElementwiseOperation b1_element_op,
+                        CElementwiseOperation c_element_op) = 0;
+
+    virtual std::unique_ptr<BaseInvoker> MakeInvokerPointer() = 0;
+};
+
+template <typename ALayout,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_batched_gemm_softmax_gemm_permute.hpp:57]: DeviceOpPtr is no longer needed

File: include/ck/tensor_operation/gpu/device/gemm_specialization.hpp (hunk 1 above)
Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/device/gemm_specialization.hpp:21]: There is 4 parameters M/N/K/O that are sufficient to describe Gemm+Gemm size. Gemm specialization merely follows that established convention. For some level of consistency, we're left with two options:
1. MNK size for Gemm0 and Gemm1, MNK padding for Gemm0 and Gemm1
2. MNKO size, MNKO padding (current)

File: include/ck/tensor_operation/gpu/device/device_batched_gemm_softmax_gemm_permute.hpp (hunk 2 above)
Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/device/device_batched_gemm_softmax_gemm_permute.hpp]: Removed in https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/370/commits/5a4a9cf16464e048749a8ad961437e4b02b7fb6e

**Review by asroy — APPROVED** [review]

**Pull request merged**